### DISCUSSION TITLE
Help! Cordova fails building iOS game

### DISCUSSION THREAD

**Discussion by konguele** [apache/cordova#441] [Q&A]
Hello!

I have created a game with Gdevelop. It works perfectly on Windows, MacOS and Android, but when I want to compile it to send it to Xcode and create the iOS version I get the following error:

```
juanjosegarciamanzano@macbook-pro-de-juan Jabato % cordova platform add ios
Using cordova-fetch for cordova-ios
Adding ios project...
Creating Cordova project for the iOS platform:
    Path: platforms/ios
    Package: com.stilugames.jabaliespacial
    Name: Jabali Espacial


[!] CocoaPods could not find compatible versions for pod "GoogleUserMessagingPlatform":
  In snapshot (Podfile.lock):
    GoogleUserMessagingPlatform (= 2.1.0, >= 1.1)

  In Podfile:
    GoogleUserMessagingPlatform (~> 2.0.0)

Specs satisfying the `GoogleUserMessagingPlatform (~> 2.0.0), GoogleUserMessagingPlatform (= 2.1.0, >= 1.1)` dependency were found, but they required a higher minimum deployment target.
```

I have tried a solution that I found on the internet that in principle seems to regenerate the framework:

```
juanjosegarciamanzano@macbook-pro-de-juan ios % pod install --repo-update
Updating local specs repositories
Analyzing dependencies
Downloading dependencies
Installing GoogleUserMessagingPlatform 2.0.1 (was 2.1.0)
Generating Pods project
Integrating client project
Pod installation complete! There are 2 dependencies from the Podfile and 6 total pods installed.

[!] The `Jabali Espacial [Debug]` target overrides the `LD_RUNPATH_SEARCH_PATHS` build setting defined in `Pods/Target Support Files/Pods-Jabali Espacial/Pods-Jabali Espacial.debug.xcconfig'. This can lead to problems with the CocoaPods installation
    - Use the `$(inherited)` flag, or
    - Remove the build settings from the target.

[!] The `Jabali Espacial [Release]` target overrides the `LD_RUNPATH_SEARCH_PATHS` build setting defined in `Pods/Target Support Files/Pods-Jabali Espacial/Pods-Jabali Espacial.release.xcconfig'. This can lead to problems with the CocoaPods installation
    - Use the `$(inherited)` flag, or
    - Remove the build settings from the target.
juanjosegarciamanzano@macbook-pro-de-juan ios % 
```
Once this is done, the following command should work, but I get this error:

```
juanjosegarciamanzano@macbook-pro-de-juan Jabato % cordova build ios
Discovered plugin "cordova-plugin-purchase". Adding it to the project
Installing "cordova-plugin-purchase" for ios

[!] The `Jabali Espacial [Debug]` target overrides the `LD_RUNPATH_SEARCH_PATHS` build setting defined in `Pods/Target Support Files/Pods-Jabali Espacial/Pods-Jabali Espacial.debug.xcconfig'. This can lead to problems with the CocoaPods installation

[!] The `Jabali Espacial [Release]` target overrides the `LD_RUNPATH_SEARCH_PATHS` build setting defined in `Pods/Target Support Files/Pods-Jabali Espacial/Pods-Jabali Espacial.release.xcconfig'. This can lead to problems with the CocoaPods installation
Command failed with exit code 1: xcodebuild -version
xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
juanjosegarciamanzano@macbook-pro-de-juan Jabato % 
```
Could you please help me? I don't know how to solve it.

Thank you!

**Comment by breautek** [comment]
There appears to be several issues...

## Deployment Target

> Specs satisfying the `GoogleUserMessagingPlatform (~> 2.0.0), GoogleUserMessagingPlatform (= 2.1.0, >= 1.1)` dependency were found, but they required a higher minimum deployment target.

This means that GoogleUserMessagingPlatform declares a higher minimum deployment target than the App project. By default cordova uses iOS 11 deployment target, which means you intend to support iOS 11 and later. But if you have a framework like GoogleUserMessagingPlatform that defines a higher deployment target (say 13 for example) then it will this error because if a framework doesn't support iOS 11, how can your app support iOS 11?

To address this, you'll have to find out the actual deployment target of GoogleUserMessagingPlatform, then you can use Cordova's `deployment-target` preference to bump to be >= GoogleUserMessagingPlatform deployment target.

As an example, this is what you would have in your `config.xml` file:

```xml
<preference name="deployment-target" value="13.0" />
```

## LD_RUNPATH_SEARCH_PATHS

> !] The `Jabali Espacial [Debug]` target overrides the `LD_RUNPATH_SEARCH_PATHS` build setting defined in `Pods/Target Support Files/Pods-Jabali Espacial/Pods-Jabali Espacial.debug.xcconfig'. This can lead to problems with the CocoaPods installation
    - Use the `$(inherited)` flag, or
    - Remove the build settings from the target.

This is telling you have your build target (presumably your app) overrides `LD_RUNPATH_SEARCH_PATHS` build setting. This is either because you're manipulating this config in a conflicting way inside your `build.json` file (if you have one) or a plugin that you're installing manipulating this config in a conflicting manner. It's hard to say with the current information available.

The solution is to use `$(inherited)` like it says so that it will inherit build configurations from higher level targets, such as project settings. But the solution might need to be applied to a plugin.

## xcode-select: error: tool 'xcodebuild' requires Xcode,

This seems to suggest that you don't have XCode installed, just the xcode command line tools. You'll need to install XCode from the App Store.

**Comment by konguele** [comment]
Hello,

Thank you for your help @breautek . We can start with the first problem. I have been trying to apply the solution you indicate, but I see in my config.xml I don't have the parameter you say. I have also looked in the index.html because it appears as source, but I have not seen the parameter either. I have attached the complete log after executing "cordova platform add iOS" command and also the config.xml, but as GitHub did not let me use that format I have moved it to txt. In case it can help.

Specs satisfying the GoogleUserMessagingPlatform (~> 2.0.0), GoogleUserMessagingPlatform (= 2.1.0, >= 1.1) dependency were found, but they required a higher minimum deployment target.


[cordova.log](https://github.com/apache/cordova/files/12909161/cordova.log)
[config.xml](https://github.com/apache/cordova/files/12909164/config.txt)

**Reply by breautek** [reply]
The `deployment-target` is a preference that *can* be found in your `config.xml` but is not required. If it's not explicitly set then it simply defaults to the cordova-ios version which is currently `11.0` as of cordova-ios@6.

So you can add it anywheres inside your `widget` tag:

```xml
<widget ...>
...
<preference name="deployment-target" value="..." />
</widget>
```

The deployment target signals the lowest iOS version you intend to support, but it must be set to at least the highest deployment target declared by your third-party libraries. (For example, you can't use deployment target 11.0 if you include a framework that itself declares deployment target 12.0).

[GoogleUserMessagingPlatform@2.1.0](https://github.com/CocoaPods/Specs/blob/master/Specs/7/e/f/GoogleUserMessagingPlatform/2.1.0/GoogleUserMessagingPlatform.podspec.json#L15) declares a minimum deployment target of `11.0`, which kind of suggesting that maybe you're using an older cordova-ios version, or perhaps there is an installed plugin that is manipulating `deployment-target`.

If you're not using cordova-ios@7 then I'd recommend updating first:

```
cordova platform remove ios
cordova platform add ios@7
```

If the issue persists, then you can try explicitly setting `deployment-target` to `11.0` (or `12.0` if you're using XCode 15, since `11.0` is technically out of support. It will still work but it will cause warnings in XCode):

```xml
<preference name="deployment-target" value="12.0" />
```

**Comment by konguele** [comment]
Good Morning,

Thanks a lot for all your help. I'm getting better, now another error is appearing. I found the last problem. Ruby was on version 2.6.0 and Cocoapods was installed on version 1.8.4, I installed cocoapods version 1.13.0 via homebrew and never used by Cordova.
I deleted everything. I upgraded Ruby to 3.2.2, I was able to install cocoapods normal, without using homebrew on version 1.13.0 and now I get this error:

Failed to install 'admob-plus-cordova': Error: Command failed with exit code 1: pod --version
/usr/local/lib/ruby/gems/3.2.0/gems/activesupport-7.1.1/lib/active_support/core_ext/array/conversions.rb:108:in `<class:Array>': undefined method `deprecator' for ActiveSupport:Module (NoMethodError)

  deprecate to_default_s: :to_s, deprecator: ActiveSupport.deprecator
                                                          ^^^^^^^^^^^
Did you mean?  deprecate_constant
	from /usr/local/lib/ruby/gems/3.2.0/gems/activesupport-7.1.1/lib/active_support/core_ext/array/conversions.rb:8:in `<top (required)>'
	from <internal:/usr/local/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
	from <internal:/usr/local/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
	from /usr/local/lib/ruby/gems/3.2.0/gems/cocoapods-1.13.0/lib/cocoapods.rb:9:in `<top (required)>'
	from <internal:/usr/local/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
	from <internal:/usr/local/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
	from /usr/local/lib/ruby/gems/3.2.0/gems/cocoapods-1.13.0/bin/pod:36:in `<top (required)>'
	from /usr/local/bin/pod:25:in `load'
	from /usr/local/bin/pod:25:in `<main>'
    at makeError (/Users/juanjosegarciamanzano/Downloads/Jabali/node_modules/execa/lib/error.js:60:11)
    at handlePromise (/Users/juanjosegarciamanzano/Downloads/Jabali/node_modules/execa/index.js:118:26)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
Failed to restore plugin "admob-plus-cordova". You might need to try adding it again. Error: Error: Command failed with exit code 1: pod --version
/usr/local/lib/ruby/gems/3.2.0/gems/activesupport-7.1.1/lib/active_support/core_ext/array/conversions.rb:108:in `<class:Array>': undefined method `deprecator' for ActiveSupport:Module (NoMethodError)

  deprecate to_default_s: :to_s, deprecator: ActiveSupport.deprecator

**Reply by breautek** [reply]
I'm not familiar with the Ruby ecosystem but I do not believe they follow semver version rules (e.g. breaking changes must be in major version increments).

The reason is because [this blog post](https://blog.cocoapods.org/CocoaPods-1.10.0-beta/) states CocoaPods dropped support for Ruby 2.0 and adds support for Ruby 2.7. And [this blog post](https://blog.cocoapods.org/CocoaPods-1.11.0-beta/) for CocoaPods 1.11 states it drops support for anything less than Ruby 2.6 and adds support for Ruby 3.0.

Because CocoaPods adds support for minor version increments, this leads to me to believe that those minor version bumps are significant and could include breaking changes and nothing states that CocoaPods supports Ruby 3.2 (however they are missing a release blog for CocoaPods 1.12 and 1.13).

> pod --version

The fact that it's simply running `pod --version` and erroring further suggest an environment incompatibility issue so I don't think CocoaPods supports Ruby 3.2. My Mac ships with Ruby 2.6 and that works out of the box.

**Comment by konguele** [comment]
Good Morning @breautek ,

I have tried different solutions, I have installed and uninstalled both ruby, cordova, cocoapods... I can't get it to work. Do you think I could send you the complete game exported from Gdevelop so you can try to reproduce the error? Maybe then we can find the solution, because I don't know what else to try.

Thank you very much!
